include/boost/url/bnf/parse.hpp

```diff
@@ -16,58 +16,79 @@
+
+   A BNF element:
+
+    - Is default constructible
+    - Holds the result, usually as a reference, of parsing
+    - Has a nested value_type
+    - Has `value_type& operator*() const noexcept`
+
+   Users should define a free function `parse` as a customization point defining how to parse their
+   BNF elements as part of the same architecture that might include arbitrary BNF elements in expressions.
+
+   These new function overloads may be defined in other namespaces. As with `std::swap`, the design relies
+   on Argument Dependent Lookup (ADL) to find these overloads.
+
+*/
+
+
 namespace boost {
 namespace urls {
 namespace bnf {
 
-/*  Design
 
-    - An element is default constructible
-    - Holds the result of parsing
-    - overloads the free function `parse`
-    - has a nested value_type
-    - has `value_type& operator*() const noexcept`
-*/
-
-/** @brief Parse a literal character
+/** Parse a substring as a BNF literal character
 
    This function parses the string defined by the first chars in the
-   range of chars [@ref it, @ref end) as a literal character.
+   range of chars [`it`, `end`) as a literal character.
 
-   The function returns `true` and resets @ref ec when the input is
-   parsed successfully. The initial character in the range is immediately
-   consumed and ignored.
-
-   If the input cannot be parsed into the literal BNF literal,
-   the function returns `false` and sets the error number in @ref ec to the
-   appropriate error number defined in @ref boost::urls::error.
+   The initial character in the range is immediately consumed and ignored.
 
    @par Example
+
    @code
    std::string str = "Hello.cpp";
    auto it = str.data();
    auto const end = it + str.size();
    if (bnf::parse(it, end, ec, 'H')) {
-       std::cout << "str begins with 'H'" << std::endl;
+       std::cout << "str begins with 'H'\n";
    }
    @endcode
 
    @par Exception Safety
-   Calls to other `parse` overloads may throw, although they should
-   prefer setting @ref ec instead.
 
-   @return `true` if the first char in the range matches @ref ch successfully,
-   `false` otherwise
+   Throws nothing
 
-   @param[in,out] it An iterator to the first element in the range.
-   At the end of this function, @ref it points to one past the last
+   @return `true` if the first char in the range matches `ch` successfully
+
+   @param it An iterator to the first element in the range.
+   At the end of this function, `it` points to one past the last
    element parsed in the range.
 
-   @param[in] end An iterator to one past the last element in the range
+   @param end An iterator to one past the last element in the range
+
+   @param ec Set to the error, if any occurred
+
+   @param ch A literal character
+
+   @see
 
-   @param[out] ec The error code this function should set @ref if it fails
+   @ref error
+
+   @par References
+
+   [Backus–Naur form](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form)
 
```

Comment:
Where's the `@li`?

Reply:
Do you always use @li for references and @see or just for references?

Reply:
OK. Just checked it. Nevermind.

Reply:
I use `<a href` but either way you need to put in parenthesis where the user is going, i.e.:
```
<a href="https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form">Backus–Naur form (Wikipedia)</a>
```

The problem though is that "BNF" doesn't really say anything. The library's API doesn't use BNF notation, nor do any of the algorithms allow you to specify BNF directly. I just use the abbreviation as a sort of proper noun to define the concept, even though actual BNF syntax is not used. You probably should never link to BNF definition anywhere in the javadocs. However you would want to link to it in the exposition.

Reply:
> The problem though is that "BNF" doesn't really say anything. 

I'm not sure I understand what you mean by that.

> The library's API doesn't use BNF notation, nor do any of the algorithms allow you to specify BNF directly. 

By "specify BNF directly" do you mean specifying it with canonical notation strings, right?

> I just use the abbreviation as a sort of proper noun to define the concept, even though actual BNF syntax is not used. 

The library is correct to use "bnf". BNF represents much more than the canonical string notation for the derivation rules, which is just a notation.

The canonical string notation for the derivation rules is not used because it's not very efficient or even worthwhile, but the whole metasyntax for the grammar and everything one needs for Chomsky's context-free grammar derivation rules are implemented by the library. 

Implementing a BNF syntax string parser from what we already have wouldn't even increment the library by much. It's just a matter of another recursive bnf element implementing the rule ` <symbol> ::= __expression__`.   

In fact, last time we talked about this a few months ago, I think my opinion was that, by looking at the reference, the library seemed to focus more on the BNF stuff than expected and that seemed to be out of scope.

> You probably should never link to BNF definition anywhere in the javadocs. However you would want to link to it in the exposition

OK. Done. I included that link because it seemed to be the most appropriate place for it, the core of the BNF parsing functionality, but I'm not attached to that link in any way. As a parallel discussion though, if the idea is that this library has nothing to do with BNF because it doesn't use the BNF syntax, this is probably not going to convince users as it is at the moment.

Reply:
I just call these types "bnfs" because I couldn't think of a better name. We could change this of course.

Reply:
You are correct to call them BNF. This is a significant part of the library. It might be dangerous to treat that as a smaller part of the library or only an informal implementation detail, given how much of the public API it represents in https://master.url.cpp.al/url/ref.html unless it were really only implementation.